### PR TITLE
Add OwningSymbol and Compilation properties to OperationAnalysisContext and SyntaxNodeAnalysisContext

### PR DESCRIPTION
**Pull request by JohnHamby**
There have been a few requests for these properties in OperationAnalysisContext. Not also adding them to SyntaxNodeAnalysisContext seemed asymmetrical.

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 1]
```diff
@@ -965,6 +977,11 @@ public SyntaxNodeAnalysisContext(SyntaxNode node, SemanticModel semanticModel, A
             _cancellationToken = cancellationToken;
         }
 
+        public SyntaxNodeAnalysisContext(SyntaxNode node, SemanticModel semanticModel, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, CancellationToken cancellationToken)
```
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: This signature is maintained because it's part of shipped public API.

**Comment by JohnHamby** [comment]
Forgot to ping some likely reviewers: @srivatsn @mavasani @genlu @nguerrera @heejaechang

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs [hunk 2]
```diff
@@ -1185,6 +1185,37 @@ class C
...
+    }
+
+    public void FunkyMethod()
+    {
+        int x = 0;
+        int y = x;
+    }
+
+    public int FunkyField = 12;
+    public int UnFunkyField = 12;
```
Comment (genlu) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs]: I understand why this makes sense in the context, but it is a little weird to think a literal (or any expression really) is owned by a field.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 3]
```diff
@@ -938,26 +939,37 @@ public struct SyntaxNodeAnalysisContext
...
         /// <summary>
         /// <see cref="CodeAnalysis.SemanticModel"/> that can provide semantic information about the <see cref="SyntaxNode"/>.
         /// </summary>
-        public SemanticModel SemanticModel { get { return _semanticModel; } }
+        public SemanticModel SemanticModel => _semanticModel;
+
+        /// <summary>
+        /// <see cref="CodeAnalysis.Compilation"/> containing the <see cref="SyntaxNode"/>.
+        /// </summary>
+        public Compilation Compilation => _semanticModel?.Compilation;
```
Comment (genlu) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Just curious, in what case _semanticModel can be null?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 4]
```diff
@@ -938,26 +939,37 @@ public struct SyntaxNodeAnalysisContext
...
+        public AnalyzerOptions Options => _options;
 
         /// <summary>
         /// Token to check for requested cancellation of the analysis.
         /// </summary>
-        public CancellationToken CancellationToken { get { return _cancellationToken; } }
+        public CancellationToken CancellationToken => _cancellationToken;
 
-        public SyntaxNodeAnalysisContext(SyntaxNode node, SemanticModel semanticModel, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, CancellationToken cancellationToken)
+        public SyntaxNodeAnalysisContext(SyntaxNode node, ISymbol owningSymbol, SemanticModel semanticModel, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, CancellationToken cancellationToken)
```
Comment (nguerrera) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: The Framework Design Guidelines require new overloads to add parameters at the end, which results in a better IntelliSense experience, IMHO.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 1 above)
Comment (nguerrera) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Should it be [Obsolete]?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 4 above)
Comment (nguerrera) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Less concerned about this if the other overload is obsolete.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Can we instead make the new API just internal? Existing public API should suffice for testability (reason we made it public in the first place)?

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs]: "OwningSymbol" seems to make sense for context objects analyzing entire code blocks or operation blocks. Probably "ContainingSymbol" seems more intuitive for syntax node and operation analysis contexts?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 5]
```diff
@@ -1007,17 +1036,17 @@ public struct OperationAnalysisContext
...
-            : this(operation, options, reportDiagnostic, isSupportedDiagnostic, semanticModel: null, cancellationToken: cancellationToken)
+        public OperationAnalysisContext(IOperation operation, ISymbol owningSymbol, Compilation compilation, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, CancellationToken cancellationToken)
+            : this(operation, owningSymbol, compilation, options, reportDiagnostic, isSupportedDiagnostic, semanticModel: null, cancellationToken: cancellationToken)
         {
         }
 
-        internal OperationAnalysisContext(IOperation operation, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, SemanticModel semanticModel, CancellationToken cancellationToken)
+        internal OperationAnalysisContext(IOperation operation, ISymbol owningSymbol, Compilation compilation, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, SemanticModel semanticModel, CancellationToken cancellationToken)
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Seems a bit odd to have 2 separate parameters for semantic model and compilation - I guess we don't have an alternative as semantic model can be null when we start raising IL to Operations?

**Comment by mavasani** [comment]
LGTM

Side note: Signatures of constructors of Context objects seem to be getting too confusing - I think we should allow breaking changes for such public APIs, which are public only for testability purposes.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 1 above)
Comment (srivatsn) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Well it's a public api whose primary use is testing but we don't know how else it's being used. I don't think adding a new constructor here is painful enough to warrant a break.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs (hunk 2 above)
Comment (srivatsn) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs]: +1 to calling it ContainingSymbol - that'll be consistent with `semanticModel.GetSymbol(node).Symbol.ContainingSymbol` which is how one would get at it today.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 3 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: I won't be null for contexts we create. However, and unfortunately, the type has a public constructor and we don't require that it be given a non-null model.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 5 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: I just realized that the semantic model is no longer necessary--it's not exposed, and the only reason we kept it internally was to get at the compilation. I'll remove the semantic model.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs (hunk 2 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs]: OK--I'll switch the name to "ContainingSymbol" for the individual node contexts. If we hadn't already shipped code block contexts with an OwningSymbol property, we might discuss changing that name too, but we did, so we won't.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 4 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: We also have a guideline that cancellation token parameters should be last. Which guideline should win?

Comment (nguerrera) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: I don't know. @terrajobst @KrzysztofCwalina -- thoughts?

Comment (terrajobst) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: The idea of putting things at the end is that a developer can easily tell which components are optional. The basic thinking being that if another overload is a prefix, then all remaining arguments are optional.

If you have multiple groups of overloads, it helps to apply this to each of them. For instance, say you've a parse method where one set of overloads parses `string` and the others are for passing `byte[]`. In that case, you wouldn't add the new overload by adding to the existing one -- you'd create a new group.

I think in your case there is no way to win. If you want to keep certain things front and certain things last, adding additional overloads will result in a mess regardless.

For this case, I think adding the overload like proposed makes the most sense.

**Comment by genlu** [comment]
:+1:

**Comment by genlu** [comment]
@JohnHamby I talked to @srivatsn about implementing a `Parent` or `Ancestors` method for operations, it seems that it would be best if `OperationAnalysisContext` provides `SemanticModel` (i.e. get the root operation from `SemanticModel` by walk up the syntax tree, then traverse down the operation tree using `OperationWalker`.) Right now `Compilation` is available to us, but it would be uber expensive to get the parent operation thought it: a quick look shows that it would require rebinding every time we call the method (this is how SemanticModel does it to get the operation for syntax node, but by providing semantic model by analysis context, those info would be cached instead of creating a new `SemanticModel` from scratch each time). Thoughts?

**Pull request merged**